cni-plugin/deployment/scripts/install-cni.sh

```diff
@@ -168,53 +192,79 @@ fi
-extension="${filename##*.}"
-if [ "${filename}" != '01-linkerd-cni.conf' ] && [ "${extension}" = 'conf' ]; then
-  echo "Renaming ${CNI_CONF_PATH} extension to .conflist"
-  CNI_CONF_PATH="${CNI_CONF_PATH}list"
-fi
+install_cni_conf() {
+  create_cni_config
+  # Things might have changed since the last time we set the variable (e.g new
+  # CNI plugin installed) so find the relevant conf path again.
+  #CNI_OLD_CONF_PATH="${CNI_OLD_CONF_PATH:-${CNI_CONF_PATH}}"
+  CNI_OLD_CONF_PATH="${CNI_CONF_PATH}"
+  CNI_CONF_PATH=$(find "${CONTAINER_MOUNT_PREFIX}${DEST_CNI_NET_DIR}" -maxdepth 1 -type f \( -iname '*conflist' -o -iname '*conf' \) | grep -v "linkerd" | sort | head -n 1)
+  # If no files have been found, create our own.
+  CNI_CONF_PATH=${CNI_CONF_PATH:-"${CONTAINER_MOUNT_PREFIX}${DEST_CNI_NET_DIR}/01-linkerd-cni.conf"}
+  CNI_CONF_FILE="${CNI_CONF_PATH}"
+  if [ -e "${CNI_CONF_FILE}" ]; then
+    # Add the linkerd-cni plugin to the existing list
+    CNI_TMP_CONF_DATA=$(cat "${TMP_CONF}")
+    CNI_CONF_DATA=$(jq --argjson CNI_TMP_CONF_DATA "$CNI_TMP_CONF_DATA" -f /linkerd/filter.jq "${CNI_CONF_FILE}")
+    echo "${CNI_CONF_DATA}" > ${TMP_CONF}
+  fi
 
-# Delete old CNI config files for upgrades.
-if [ "${CNI_CONF_PATH}" != "${CNI_OLD_CONF_PATH}" ]; then
-  echo "Removing CNI_OLD_CONF_PATH: ${CNI_OLD_CONF_PATH}"
-  rm -f "${CNI_OLD_CONF_PATH}"
-fi
+  # If the old config filename ends with .conf, rename it to .conflist, because it has changed to be a list
+  filename=${CNI_CONF_PATH##*/}
+  extension="${filename##*.}"
+  if [ "${filename}" != '01-linkerd-cni.conf' ] && [ "${extension}" = 'conf' ]; then
+    echo "Renaming ${CNI_CONF_PATH} extension to .conflist"
+    CNI_CONF_PATH="${CNI_CONF_PATH}list"
+  fi
+
+  # Delete old CNI config files for upgrades.
+  if [ "${CNI_CONF_PATH}" != "${CNI_OLD_CONF_PATH}" ]; then
+    echo "Removing CNI_OLD_CONF_PATH: ${CNI_OLD_CONF_PATH}"
+    rm -f "${CNI_OLD_CONF_PATH}"
+  fi
 
-# Move the temporary CNI config into place.
-mv "${TMP_CONF}" "${CNI_CONF_PATH}" || exit_with_error 'Failed to mv files.'
+  # Move the temporary CNI config into place.
+  mv "${TMP_CONF}" "${CNI_CONF_PATH}" || exit_with_error 'Failed to mv files.'
 
-echo "Created CNI config ${CNI_CONF_PATH}"
+  echo "Created CNI config ${CNI_CONF_PATH}"
+
+}
 
-# Unless told otherwise, sleep forever.
-# This prevents Kubernetes from restarting the pod repeatedly.
-should_sleep=${SLEEP:-"true"}
-echo "Done configuring CNI. Sleep=$should_sleep"
-while [ "${should_sleep}" = 'true'  ]; do
-  sleep infinity &
-  wait $!
-done
+install_cni_bin
+
+install_cni_conf
+
+# Start looping and watching fs events
+HOST_CNI_NET="${CONTAINER_MOUNT_PREFIX}${DEST_CNI_NET_DIR}"
+inotifywait -m ${HOST_CNI_NET} -e moved_to -e create -e delete |
+ while read -r directory action filename; do
+   if [[ "$filename" =~ .*.(conflist|conf) ]]; then 
+    filepath="$directory$filename"
+    if [ "$action" = "DELETE" ] && [ $filepath = "${CNI_CONF_PATH}" ]; then
```

Comment:
do we care about handling moved_from events?

Reply:
In general, it feels a little sketchy to me that this block uses variables (`CNI_CONF_PATH`, etc) that are re-defined  every time `install_cni_conf` is invoked...

Reply:
re: `move_to` and `move_from` @olix0r 

They've been a bit of a pain to deal with, here's how I understand it, maybe you can confirm whether this makes sense.

I think when the files are on the same partition, we see the events properly (i.e as a `MOVE_TO`); that's because only the name of the entry/inode gets changed. If src - tgt are not on the same partition, we do a `cp` followed by a `rm` on the src.

In the CNI plugin container, we only ever see `CREATE` and `DELETE` events, despite actually moving stuff. I exec'd on the container and tried it out:

```text
in the temporary directory where we create the config, we see the behaviour we'd expect:
/tmp/ CREATE sed9a4PSq
/tmp/ MOVED_TO linkerd-cni.conf.default
```

We create a tmp file and then "`mv`" it under a different name in the same directory. I was confused why here I was seeing a `MOVED_TO`, but not on the `"hostPath"`. I created a file in the tmp dir and tried to move it on the host, while an inotifywait watch was running on the `"hostPath"`:

```sh
root@k3d-cni-server-0:/tmp# mv banan.sh /host/etc/cni/net.d

root@k3d-cni-server-0:/host/etc/cni/net.d# inotifywait -m $PWD -e moved_to,create,delete
Setting up watches.
Watches established.

/host/etc/cni/net.d/ CREATE banan.sh
```

I assume they'd be on different partitions and that's why we don't get a move but a create (i.e an underlying cp from src to dst in the mv cmd)? Inotifywait's docs aren't very helpful here...

```text
moved_to
A file or directory was moved into a watched directory. This event occurs even if the file is simply moved from and to the same directory.
```

From `man mv`

```text
 Rename SOURCE to DEST, or move SOURCE(s) to DIRECTORY.
```

I can't find the exact quote about the partition now, not even in the [docs](https://www.gnu.org/software/coreutils/manual/html_node/mv-invocation.html) but it would explain I guess why I'm never seeing `moved_to` events.